Add a basic TLS description (#303)
owncloud/docs-ocis: modules/ROOT/pages/deployment/services/tls.adoc

= Transport Layer Security (TLS)
:toc: macro

:description: TLS is important to secure communication between parties. Infinite Scale provides TLS for internal but also for communication with users or services which are not provided by ocis. 

include::partial$deployment/services/beta-statement.adoc[]

// remove the remark to get the toc rendered if there are more than one headlines
{empty} +

toc::[]

== Introduction

> {description} While focussing on describing the environment variables and relationship, managing certificates is not in the scope of this document.

Reviewer: ```suggestion
{description} While focusing on describing the environment variables and their relationships, managing certificates is not the scope of this document.
```

> If not explicitely mentioned, you will find the environmen variables in service starting with the environment variable name. Names starting with `OCIS_` generally have a global scope and are not bound to a particular service. 

Reviewer: Please check if this suggestion is correct.

== Secure Service-To-Service Communication

When deploying a distributed setup, it is recommended to secure communication and authentication between services.

Most services currently provide TLS for service-to-service communication and authentication with the exception of the xref:{s-path}/frontend.adoc[frontend] and the xref:{s-path}/ocdav.adoc[ocdav] service. TLS will be added to the latter ones in a future release.

> NOTE: TLS is usually defined per instance on a global scope, though you can define the settings per service too.

Author: ```suggestion
[NOTE]
====
* TLS is usually defined per instance on a global scope, though you can define the settings per service too.
* When setting TLS modes to _insecure_, the traffic can't be sniffed anymore, but it would be still possible to intercept it.
====
```

=== TLS for gRPC

OCIS_GRPC_TLS_ENABLED::
+
--
Activates TLS for the gRPCS-based services using the server certificate and key configured via: 

* `OCIS_GRPC_TLS_CERTIFICATE` and
* `OCIS_GRPC_TLS_KEY`.

> If `OCIS_GRPC_TLS_CERTIFICATE` is not set, a temporary server certificate is generated - to be used with `OCIS_GRPC_CLIENT_TLS_MODE`=insecure.

Reviewer 2: yes. let us add this

> If `OCIS_GRPC_TLS_CERTIFICATE` is not set, a temporary server certificate is generated - to be used with `OCIS_GRPC_CLIENT_TLS_MODE`=insecure.

Author: Added as general note in line 28.

--

OCIS_GRPC_CLIENT_TLS_MODE::
+
--
TLS mode for gRPC connection to the go-micro based gRPC services. Possible values are `off`, `insecure` and `on`.

* `off`: disables transport security for the clients. 

> * `insecure` allows to use transport security, but disables certificate verification (to be used with the autogenerated self-signed certificates).

Reviewer 3: ```suggestion
* `insecure` allows transport security to be used, but disables certificate verification (to be used with the autogenerated self-signed certificates).
```

* `on` enables transport security, including server ceritificate verification.
--

OCIS_GRPC_TLS_KEY::
Path/File name for the TLS certificate key (in PEM format) for the server certificate to use for the gRPC services.

OCIS_GRPC_TLS_CERTIFICATE::
Path/File name of the TLS server certificate (in PEM format) for the gRPC services.

OCIS_GRPC_CLIENT_TLS_CACERT::
Path/File name for the root CA certificate used to validate TLS server certificates of the go-micro based gRPC services.

=== TLS for the HTTP Backend

OCIS_HTTP_TLS_ENABLED::
text to be added

PROXY_INSECURE_BACKENDS::
Disable TLS certificate validation for all HTTP backend connections.
+

> NOTE: This environment variable will most likely be replaced by `OCIS_INSECURE_BACKENDS` for naming harmonisation in a later release.

Reviewer: ```suggestion
NOTE: This environment variable will most likely be replaced by `OCIS_INSECURE_BACKENDS` for naming harmonization in a later release.
```

> NOTE: This environment variable will most likely be replaced by `OCIS_INSECURE_BACKENDS` for naming harmonization in a later release.

Reviewer 2: please do not do any forecasting here `NOTE: This environment variable will most likely be replaced by `OCIS_INSECURE_BACKENDS` for naming harmonization in a later release`

> NOTE: This environment variable will most likely be replaced by `OCIS_INSECURE_BACKENDS` for naming harmonization in a later release.

Author: ```suggestion
```

> NOTE: This environment variable will most likely be replaced by `OCIS_INSECURE_BACKENDS` for naming harmonization in a later release.

Author: removed

OCIS_HTTP_TLS_KEY::
Path/File name for the TLS certificate key (in PEM format) for the server certificate to use for the http services.

OCIS_HTTP_TLS_CERTIFICATE::

> Path/File name of the TLS server certificate (in PEM format) for the http services.

Reviewer: ```suggestion
Path/File name of the TLS server certificate (in PEM format) for the HTTP services.
```

// OCIS_HTTP_TLS_CACERT::
// missing

== Other TLS Configurations

> These TLS configurations are used for securing communication to users or services not provided by Infinite Scale.

Author: Will add

=== TLS for LDAP

You will find the LDAP TLS settings in the xref:{s-path}/idp.adoc[IDP] and xref:{s-path}/auth-basic.adoc[Auth-Basic] service.

LDAP_CACERT::
Path to the TLS certificate for the LDAP service.

LDAP_INSECURE::

> Allow insecure connections to the user backend like LDAP, CS3 api, …​

Reviewer 3: ```suggestion
Allow insecure connections to the user backend like LDAP, CS3 API, …​
```

API is an acronym. I suppose it should always be in uppercase.

* `on` enables transport security, including server ceritificate verification.
--

OCIS_GRPC_TLS_KEY::
Path/File name for the TLS certificate key (in PEM format) for the server certificate to use for the gRPC services.

OCIS_GRPC_TLS_CERTIFICATE::
Path/File name of the TLS server certificate (in PEM format) for the gRPC services.

OCIS_GRPC_CLIENT_TLS_CACERT::
Path/File name for the root CA certificate used to validate TLS server certificates of the go-micro based gRPC services.

=== TLS for the HTTP Backend

OCIS_HTTP_TLS_ENABLED::
Activates TLS for the http based services using the server certifcate and key configured via OCIS_HTTP_TLS_CERTIFICATE and OCIS_HTTP_TLS_KEY. If OCIS_HTTP_TLS_CERTIFICATE is not set a temporary server certificate is generated - to be used with PROXY_INSECURE_BACKEND=true.

PROXY_INSECURE_BACKENDS::
Disable TLS certificate validation for all HTTP backend connections.

OCIS_HTTP_TLS_KEY::
Path/File name for the TLS certificate key (in PEM format) for the server certificate to use for the HTTP services.

OCIS_HTTP_TLS_CERTIFICATE::
Path/File name of the TLS server certificate (in PEM format) for the HTTP services.

PROXY_HTTPS_CACERT::

> The root CA certificate used to validate TLS server certificates of HTTPS enabled backend services.

Author: ```suggestion
Path/File for the root CA certificate used to validate the server’s TLS certificate for https enabled backend services.
```

=== TLS for the Event Bus (NATS)

OCIS_EVENTS_ENABLE_TLS::
Enable TLS for the connection to the events broker. The events broker is the ocis service which receives and delivers events between services.

OCIS_INSECURE::
Whether the server should skip the client certificate verification during the TLS handshake.

> NATS_TLS_CERT::
> Path/File name of the TLS server certificate for the nats listener.
> 
> NATS_TLS_KEY::
> Path/File name for the TLS certificate key for the server certificate.

Author: ```suggestion
NATS_TLS_CERT::
Path/File name of the TLS server certificate (in PEM format) for the NATS listener.

NATS_TLS_KEY::
Path/File name for the TLS certificate key (in PEM format) for the NATS listener.
```

OCIS_HTTP_TLS_CERTIFICATE::
Path/File name of the TLS server certificate (in PEM format) for the http services.

OCIS_HTTP_TLS_KEY::
Path/File name for the TLS certificate key (in PEM format) for the server certificate to use for the http services.

OCIS_GRPC_CLIENT_TLS_MODE::
TLS mode for grpc connection to the go-micro based grpc services. Possible values are 'off', 'insecure' and 'on'. 'off': disables transport security for the clients. 'insecure' allows to use transport security, but disables certificate verification (to be used with the autogenerated self-signed certificates). 'on' enables transport security, including server ceritificate verification.

OCIS_GRPC_CLIENT_TLS_CACERT::
Path/File name for the root CA certificate (in PEM format) used to validate TLS server certificates of the go-micro based grpc services.

== Other TLS Configurations

These TLS configurations are used for securing communication to users or services not provided by Infinite Scale.

=== TLS for SMTP

NOTIFICATIONS_SMTP_ENCRYPTION::
Encryption method for the SMTP communication. Possible values are `starttls`, `ssl`, `ssltls`, 'tls` and `none`.

=== TLS for WOPI

APP_PROVIDER_WOPI_INSECURE::

> Allow insecure connections to the app.

Author: ```suggestion
Disable TLS certificate validation for requests to the WOPI server and the web office application. Do not set this in production environments.
```

APP_PROVIDER_WOPI_WOPI_SERVER_IOP_SECRET::
Shared secret of the CS3org WOPI server.

=== TLS for LDAP

You will find the LDAP TLS settings in the xref:{s-path}/idp.adoc[IDP] and xref:{s-path}/auth-basic.adoc[Auth-Basic] service.

LDAP_CACERT::
Path to the TLS certificate for the LDAP service.

LDAP_INSECURE::
Disable TLS certificate validation for the LDAP connections. Do not set this in production environments.

=== TLS for the HTTP Frontend

PROXY_TLS::
Use the HTTPS server instead of the HTTP server.

PROXY_TRANSPORT_TLS_KEY::
Path/File name of the TLS server certificate key for the HTTPS server.

PROXY_TRANSPORT_TLS_CERT::
Path/File name of the TLS server certificate for the HTTPS server.

PROXY_HTTPS_CACERT::

> The root CA certificate used to validate TLS server certificates of HTTPS enabled backend services.

Author: ```suggestion
Path/File for the root CA certificate used to validate the server’s TLS certificate for https enabled backend services.
```
